File: go/0501E-二叉搜索树中的众数.go
/*
 * @Author: SukiEva
 * @Date: 2021-11-11 09:49:20
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func findMode(root *TreeNode) []int {
	ans := make([]int, 0)
	maxCount, cnt := 0, 0
	var pre *TreeNode
	var inOrder func(node *TreeNode)
	inOrder = func(node *TreeNode) {
		if node == nil {
			return
		}
		inOrder(node.Left)

		if pre == nil {
			cnt = 1
		} else if pre.Val == node.Val {
			cnt++
		} else {
			cnt = 1
		}
		pre = node

		if cnt == maxCount {
			ans = append(ans, node.Val)
		} else if cnt > maxCount {
			maxCount = cnt
			ans = []int{node.Val}
		}
		inOrder(node.Right)
	}
	inOrder(root)
	return ans
}

File: cpp/0501E-二叉搜索树中的众数.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-22 10:46:26
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    vector<int> ans;
    TreeNode* pre = nullptr;
    int cnt = 0;
    int Max = 0;
    void traversal(TreeNode* root) {
        if (root == nullptr) return;
        traversal(root->left);
        if (pre == nullptr)
            cnt = 1;
        else if (pre->val == root->val)
            cnt++;
        else
            cnt = 1;
        pre = root;

        if (cnt == Max)
            ans.push_back(root->val);
        else if (cnt > Max) {
            Max = cnt;
            ans.clear();
            ans.push_back(root->val);
        }
        traversal(root->right);
    }

    vector<int> findMode(TreeNode* root) {
        traversal(root);
        return ans;
    }
};